README.md
# d3js.example

Working with D3.js examples

### Keywords

`Data` `D3` `Data-Driven` `Documents` `Visualization` `Tools` `JavaScript`

### Documentation

A detailed list of the changes and fixes in each version update can be found in the CHANGELOG.md.

* [Change Log] (CHANGELOG.md)

### Resources

* [D3js] (http://d3js.org/) - D3 Data-Driven Documents
* [mbostock] (http://bl.ocks.org/mbostock) - examples
* [Dashing D3] (https://www.dashingd3js.com/) - Tutorials


CHANGELOG.md
# CHANGELOG for D3js.example

[![Semver](http://img.shields.io/SemVer/2.0.0.png)](http://semver.org/spec/v2.0.0.html)

Major/Minor/Patch 0.0.0

### Version 0.1.0 (05/28/2014)

js/main.js
/*! main.js | d3.js_example */

var margin = {top: 20, right: 120, bottom: 20, left: 120},
    width = 960 - margin.right - margin.left,
    height = 800 - margin.top - margin.bottom;

var i = 0,
    duration = 750,
    root;

var tree = d3.layout.tree()
    .size([height, width]);

var diagonal = d3.svg.diagonal()
    .projection(function(d) { return [d.y, d.x]; });

var svg = d3.select("body").append("svg")
    .attr("width", width + margin.right + margin.left)
    .attr("height", height + margin.top + margin.bottom)
  .append("g")
    .attr("transform", "translate(" + margin.left + "," + margin.top + ")");

// var json = getData();
//   root = json;
//   root.x0 = height / 2;
//   root.y0 = 0;

//   function collapse(d) {
//     if (d.children) {
//       d._children = d.children;
//       d._children.forEach(collapse);
//       d.children = null;
//     }
//   }

//   root.children.forEach(collapse);
//   update(root);

d3.json("data/sherwood.json", function(error, flare) {
  root = flare;
  root.x0 = height / 2;
  root.y0 = 0;

  function collapse(d) {
    if (d.children) {
      d._children = d.children;
      d._children.forEach(collapse);
      d.children = null;
    }
  }

  root.children.forEach(collapse);
  update(root);
});

d3.select(self.frameElement).style("height", "800px");

function update(source) {

  // Compute the new tree layout.
  var nodes = tree.nodes(root).reverse(),
      links = tree.links(nodes);

  // Normalize for fixed-depth.
  nodes.forEach(function(d) { d.y = d.depth * 180; });

  // Update the nodes…
  var node = svg.selectAll("g.node")
      .data(nodes, function(d) { return d.id || (d.id = ++i); });

  // Enter any new nodes at the parent's previous position.
  var nodeEnter = node.enter().append("g")
      .attr("class", "node")
      .attr("transform", function(d) { return "translate(" + source.y0 + "," + source.x0 + ")"; })
      .on("click", click);

  nodeEnter.append("circle")
      .attr("r", 1e-6)
      .style("fill", function(d) { return d._children ? "lightsteelblue" : "#fff"; });

  nodeEnter.append("text")
      .attr("x", function(d) { return d.children || d._children ? -10 : 10; })
      .attr("dy", ".35em")
      .attr("text-anchor", function(d) { return d.children || d._children ? "end" : "start"; })
      .text(function(d) { return d.name; })
      .style("fill-opacity", 1e-6);

  // Transition nodes to their new position.
  var nodeUpdate = node.transition()
      .duration(duration)
      .attr("transform", function(d) { return "translate(" + d.y + "," + d.x + ")"; });

  nodeUpdate.select("circle")
      .attr("r", 4.5)
      .style("fill", function(d) { return d._children ? "lightsteelblue" : "#fff"; });

  nodeUpdate.select("text")
      .style("fill-opacity", 1);

  // Transition exiting nodes to the parent's new position.
  var nodeExit = node.exit().transition()
      .duration(duration)
      .attr("transform", function(d) { return "translate(" + source.y + "," + source.x + ")"; })
      .remove();

  nodeExit.select("circle")
      .attr("r", 1e-6);

  nodeExit.select("text")
      .style("fill-opacity", 1e-6);

  // Update the links…
  var link = svg.selectAll("path.link")
      .data(links, function(d) { return d.target.id; });

  // Enter any new links at the parent's previous position.
  link.enter().insert("path", "g")
      .attr("class", "link")
      .attr("d", function(d) {
        var o = {x: source.x0, y: source.y0};
        return diagonal({source: o, target: o});
      });

  // Transition links to their new position.
  link.transition()
      .duration(duration)
      .attr("d", diagonal);

  // Transition exiting nodes to the parent's new position.
  link.exit().transition()
      .duration(duration)
      .attr("d", function(d) {
        var o = {x: source.x, y: source.y};
        return diagonal({source: o, target: o});
      })
      .remove();

  // Stash the old positions for transition.
  nodes.forEach(function(d) {
    d.x0 = d.x;
    d.y0 = d.y;
  });
}

// Toggle children on click.
function click(d) {
  if (d.children) {
    d._children = d.children;
    d.children = null;
  } else {
    d.children = d._children;
    d._children = null;
  }
  update(d);
}

function getData() {
    return {
"name": "Vern & Eileen",
"lastname": "Wolfley",
"spouse1": "Vern",
"spouse2": "Eileen",
"children": [
    {
    "name": "Vern & Kindy",
    "lastname": "Wolfley",
    "spouse1": "Vern",
    "spouse2": "Kindy",
    "children": [
        {"name": "Elena", "year": "", "age": "", "order": ""},
        {"name": "Natalie", "year": "", "age": "", "order": ""},
        {"name": "Meghan", "year": "", "age": "", "order": ""},
        {"name": "Vern", "year": "", "age": "", "order": ""},
        {"name": "Benjamin", "year": "", "age": "", "order": ""}
    ]
    },
    {
    "name": "Dennis & Kena",
    "lastname": "Wolfley",
    "spouse1": "Dennis",
    "spouse2": "Kena",
    "children": [
        {"name": "Baby1", "year": "", "age": "", "order": ""}
    ]
    },
    {
    "name": "Daniel & Joyce",
    "lastname": "Wolfley",
    "spouse1": "Daniel",
    "spouse2": "Joyce",
    "children": [
        {"name": "Baby1", "year": "", "age": "", "order": ""},
        {"name": "Baby2", "year": "", "age": "", "order": ""},
        {"name": "Baby3", "year": "", "age": "", "order": ""},
        {"name": "Baby4", "year": "", "age": "", "order": ""}
    ]
    },
    {
    "name": "Sara & Doug",
    "lastname": "Bodily",
    "spouse1": "Sara",
    "spouse2": "Doug",
    "children": [
        {"name": "Baby1", "year": "", "age": "", "order": ""},
        {"name": "Baby2", "year": "", "age": "", "order": ""},
        {"name": "Baby3", "year": "", "age": "", "order": ""},
        {"name": "Baby4", "year": "", "age": "", "order": ""},
        {"name": "Baby5", "year": "", "age": "", "order": ""}
    ]
    },
    {
    "name": "Anna & Alex",
    "lastname": "Ellege",
    "spouse1": "Anna",
    "spouse2": "Alex",
    "children": [
        {"name": "Baby1", "year": "", "age": "", "order": ""},
        {"name": "Baby2", "year": "", "age": "", "order": ""},
        {"name": "Baby3", "year": "", "age": "", "order": ""},
        {"name": "Baby4", "year": "", "age": "", "order": ""}
    ]
    },
    {
    "name": "Aaron & Chelsea",
    "lastname": "Wolfley",
    "spouse1": "Aaron",
    "spouse2": "Chelsea",
    "children": [
        {"name": "Baby1", "year": "", "age": "", "order": ""},
        {"name": "Baby2", "year": "", "age": "", "order": ""}
    ]
    },
    {
    "name": "Ammon & Emilee",
    "lastname": "Wolfley",
    "spouse1": "Ammon",
    "spouse2": "Emilee",
    "children": [
        {"name": "Wes", "year": "", "age": "", "order": ""},
        {"name": "Olive", "year": "", "age": "", "order": ""}
    ]
    }
]
};
}

js/main_old1.js
/*! main.js | d3.js_example */
// for use with example to change display type
//
var margin = {top: 20, right: 20, bottom: 20, left: 20};
var width = 960;
var height = 800;
var diameter = 600;
var duration = 2000;

d3.selectAll("input").on("change", change);

function change() {
    if (this.value === "radialtree")
        transitionToRadialTree();
    else if (this.value === "radialcluster")
        transitionToRadialCluster();
    else if (this.value === "tree")
        transitionToTree();
    else
        transitionToCluster();
};

function transitionToRadialTree() {

    var nodes = radialTree.nodes(root), // recalculate layout
        links = radialTree.links(nodes);

    svg.transition().duration(duration)
        .attr("transform", "translate(" + (width/2) + "," + (height/2) + ")");
        // set appropriate translation (origin in middle of svg)

    link.data(links)
        .transition()
        .duration(duration)
        .style("stroke", "#fc8d62")
        .attr("d", radialDiagonal); //get the new radial path

    node.data(nodes)
        .transition()
        .duration(duration)
        .attr("transform", function(d) {
            return "rotate(" + (d.x - 90) + ")translate(" + d.y + ")";
        });

    node.select("circle")
        .transition()
        .duration(duration)
        .style("stroke", "#984ea3");

};

function transitionToRadialCluster() {

    var nodes = radialCluster.nodes(root), // recalculate layout
        links = radialCluster.links(nodes);

    svg.transition().duration(duration)
        .attr("transform", "translate(" + (width/2) + "," +
                                          (height/2) + ")");
        // set appropriate translation (origin in middle of svg)

    link.data(links)
        .transition()
        .duration(duration)
        .style("stroke", "#66c2a5")
        .attr("d", radialDiagonal); //get the new radial path

    node.data(nodes)
        .transition()
        .duration(duration)
        .attr("transform", function(d) {
            return "rotate(" + (d.x - 90) + ")translate(" + d.y + ")";
        });

    node.select("circle")
        .transition()
        .duration(duration)
        .style("stroke", "#4daf4a");

};

function transitionToTree() {

    var nodes = tree.nodes(root), //recalculate layout
        links = tree.links(nodes);

    svg.transition().duration(duration)
        .attr("transform", "translate(40,0)");

    link.data(links)
        .transition()
        .duration(duration)
        .style("stroke", "#e78ac3")
        .attr("d", diagonal); // get the new tree path

    node.data(nodes)
        .transition()
        .duration(duration)
        .attr("transform", function (d) {
            return "translate(" + d.y + "," + d.x + ")";
        });

    node.select("circle")
        .transition()
        .duration(duration)
        .style("stroke", "#377eb8");

};

function transitionToCluster() {

    var nodes = cluster.nodes(root), //recalculate layout
        links = cluster.links(nodes);

    svg.transition().duration(duration)
        .attr("transform", "translate(40,0)");

    link.data(links)
        .transition()
        .duration(duration)
        .style("stroke", "#8da0cb")
        .attr("d", diagonal); //get the new cluster path

    node.data(nodes)
        .transition()
        .duration(duration)
        .attr("transform", function (d) {
            return "translate(" + d.y + "," + d.x + ")";
        });

    node.select("circle")
        .transition()
        .duration(duration)
        .style("stroke", "#e41a1c");

};

var root; // store data in a variable accessible by all functions

var tree = d3.layout.tree()
    .size([height, width - 160]);

var cluster = d3.layout.cluster()
    .size([height, width - 160]);

var diagonal = d3.svg.diagonal()
    .projection(function (d) {
    return [d.y, d.x];
});

var radialTree = d3.layout.tree()
    .size([360, diameter / 2 ])
    .separation(function(a, b) {
        return (a.parent == b.parent ? 1 : 2) / a.depth;
    });

var radialCluster = d3.layout.cluster()
    .size([360, diameter / 2 ])
    .separation(function(a, b) {
        return (a.parent == b.parent ? 1 : 2) / a.depth;
    });

var radialDiagonal = d3.svg.diagonal.radial()
    .projection(function(d) {
        return [d.y, d.x / 180 * Math.PI];
    });


var svg = d3.select("body").append("svg")
    .attr("width", width)
    .attr("height", height)
    .append("g")
    .attr("transform", "translate(40,0)");

    var root = getData(),
        nodes = cluster.nodes(root),
        links = cluster.links(nodes);

    var link = svg.selectAll(".link")
        .data(links)
       .enter()
        .append("path")
        .attr("class", "link")
        .style("stroke", "#8da0cb")
        .attr("d", diagonal);

    var node = svg.selectAll(".node")
        .data(nodes)
       .enter()
        .append("g")
        .attr("class", "node")
        .attr("transform", function (d) {
        return "translate(" + d.y + "," + d.x + ")";
    })

    node.append("circle")
        .attr("r", 4.5)
        .style("stroke", "#e41a1c");

    node.append("text")
        .attr("dx", function (d) { return d.children ? -8 : 8; })
        .attr("dy", 3)
        .style("text-anchor", function (d) { return d.children ? "end" : "start"; })
        .text(function (d) { return d.name; });


function getData() {
    return {
"name": "Vern & Eileen",
"lastname": "Wolfley",
"spouse1": "Vern",
"spouse2": "Eileen",
"children": [
    {
    "name": "Vern & Kindy",
    "lastname": "Wolfley",
    "spouse1": "Vern",
    "spouse2": "Kindy",
    "children": [
        {"name": "Elena", "year": "", "age": "", "order": ""},
        {"name": "Natalie", "year": "", "age": "", "order": ""},
        {"name": "Meghan", "year": "", "age": "", "order": ""},
        {"name": "Vern", "year": "", "age": "", "order": ""},
        {"name": "Benjamin", "year": "", "age": "", "order": ""}
    ]
    },
    {
    "name": "Dennis & Kena",
    "lastname": "Wolfley",
    "spouse1": "Dennis",
    "spouse2": "Kena",
    "children": [
        {"name": "Baby1", "year": "", "age": "", "order": ""}
    ]
    },
    {
    "name": "Daniel & Joyce",
    "lastname": "Wolfley",
    "spouse1": "Daniel",
    "spouse2": "Joyce",
    "children": [
        {"name": "Baby1", "year": "", "age": "", "order": ""},
        {"name": "Baby2", "year": "", "age": "", "order": ""},
        {"name": "Baby3", "year": "", "age": "", "order": ""},
        {"name": "Baby4", "year": "", "age": "", "order": ""}
    ]
    },
    {
    "name": "Sara & Doug",
    "lastname": "Bodily",
    "spouse1": "Sara",
    "spouse2": "Doug",
    "children": [
        {"name": "Baby1", "year": "", "age": "", "order": ""},
        {"name": "Baby2", "year": "", "age": "", "order": ""},
        {"name": "Baby3", "year": "", "age": "", "order": ""},
        {"name": "Baby4", "year": "", "age": "", "order": ""},
        {"name": "Baby5", "year": "", "age": "", "order": ""}
    ]
    },
    {
    "name": "Anna & Alex",
    "lastname": "Ellege",
    "spouse1": "Anna",
    "spouse2": "Alex",
    "children": [
        {"name": "Baby1", "year": "", "age": "", "order": ""},
        {"name": "Baby2", "year": "", "age": "", "order": ""},
        {"name": "Baby3", "year": "", "age": "", "order": ""},
        {"name": "Baby4", "year": "", "age": "", "order": ""}
    ]
    },
    {
    "name": "Aaron & Chelsea",
    "lastname": "Wolfley",
    "spouse1": "Aaron",
    "spouse2": "Chelsea",
    "children": [
        {"name": "Baby1", "year": "", "age": "", "order": ""},
        {"name": "Baby2", "year": "", "age": "", "order": ""}
    ]
    },
    {
    "name": "Ammon & Emilee",
    "lastname": "Wolfley",
    "spouse1": "Ammon",
    "spouse2": "Emilee",
    "children": [
        {"name": "Wes", "year": "", "age": "", "order": ""},
        {"name": "Olive", "year": "", "age": "", "order": ""}
    ]
    }
]
};
}